### PR TITLE
Hashing In Data Structure

### PR DESCRIPTION
**Pull request by anubhavb11**
### Article checklist

**ATTENTION: In your PR - add the ISSUE (topic form) # that this PR corresponds to. You can do this by also adding a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.**
      
      
**NOTE: (Please ensure that you have only one open issue + linked pull request at a time. This will ensure that we complete the article in a timely manner from inception to publishing.)**

**If you have not already please go over our [Resources Page](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md) for more tips and tricks on how to improve our overall technical writing.**

#### Preliminary Checks - Formatting and Structure 
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**
- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
- [x] In your PR - add a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) -  [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker? We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**

### Contribution guidelines
For first-time contributors and for more details, **see [Contributing Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**

### CODE REVIEW

**Comment by Bayler** [comment]
Hi @anubhavb11 Welcome to Section. I'll be reviewing your article.

Before I continue reviewing, you need to do a couple of things first:

1. Upload a hero image and name it as hero.jpg or hero.png. Also make sure it's 600 by 400px
2. Your main file needs to be named as index.md
3. On the preliminary checks, there is an instruction where they mention using header #3 and above. I've noticed you've used H1. Please change i.e., h3, h4, h5 etc
4. In your PR - add the ISSUE (topic form) # that this PR corresponds to. You can do this by also adding a comment that states "This fixes #(enter your Idea Content Form #)" to link and close your Idea Content (issue) when the PR is merged.
5. Provide a link to your code implementation on www.replit.com so that I can confirm your code and output

**Comment by anubhavb11** [comment]
https://replit.com/@Anubhavb11/HashMap#main.cpp

**Comment by Bayler** [comment]
@anubhavb11 The 4th instruction should be done on this PR and not on the issue. It should reference the issue, not vice versa

**Comment by anubhavb11** [comment]
"This fixes ##2460 "

File: content/articles/hashing-in-data-structure/index.md [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+    alt: hero image example image
+---
+### Hashing in Data Structure
+
+In hashing, the idea is to use a hash function that converts a given key to a smaller number and uses the small number as an index in a table called a hash table.
+
+We generate a hash for the input using the hash function and then store the element using the generated hash as the key in the hash table.
+
+![HashTable](./HashIntro.PNG)
+
```
Comment (Bayler) [inline content/articles/hashing-in-data-structure/index.md:24]: Are all the images quoted in this article yours? If not, you need to add the sources

File: content/articles/hashing-in-data-structure/index.md [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+#### Characteristics of a Good Hash Function :
+
+- **Uniform Distribution**: For distribution throughout the constructed table.
+- **Fast**: The generation of hash should be very fast, and should not produce any considerable overhead.
+
+#### Collision Hashing Techniques:
+
+1. **Open Hashing (Separate Chaining)**: In this using Linked List, new nodes are added to the list, the key act as the head pointer, pointing to several nodes having different values. If any collision two elements is having same hash value we store both of them in a same linked list.
+
```
Comment (Bayler) [inline content/articles/hashing-in-data-structure/index.md:57]: This sentence is hard to follow. Consider rephrasing it in a way someone can understand. Also add commas where necessary

File: content/articles/hashing-in-data-structure/index.md (hunk 1 above)
Comment (anubhavb11) [reply content/articles/hashing-in-data-structure/index.md:24]: Yes, all the images are made by me.

File: content/articles/hashing-in-data-structure/index.md [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+If the average number of items in a block exceeds the load factor, the elements are rehashed with a larger hash table size.
+
+**Rehashing**
+
+When the load factor gets "too high" (specified by the threshold value), the process of raising the size of the hash table begins, expecting that collisions would become more common.
+
+- Typically, double the size of the table (but still prime).
+- All existing items must be reinserted into the hash table.
+
```
Comment (Bayler) [inline content/articles/hashing-in-data-structure/index.md:116]: This section is not clear. Are they the solutions to the load factor getting "too high"? Please make it clear

File: content/articles/hashing-in-data-structure/index.md [hunk 4]
```diff
@@ -0,0 +1,299 @@
...
+```
+
+The complete execution of the above code can be found [here](https://replit.com/@Anubhavb11/HashMap#main.cpp).
+=======
+
+#### Output
+
+![output](./Capture.PNG)
+
```
Comment (Bayler) [inline content/articles/hashing-in-data-structure/index.md:297]: Please explain in detail these output results. Also, it's good practice to end an article with a "Conclusion" section which wraps up what your article is about.

File: content/articles/hashing-in-data-structure/index.md [hunk 5]
```diff
@@ -0,0 +1,299 @@
...
+excerpt_separator: <!--more-->
+images:
+
+  - url: /hashing-in-data-structure/hero.jpg 
+    alt: hero image example image
+---
+### Hashing in Data Structure
+
+In hashing, the idea is to use a hash function that converts a given key to a smaller number and uses the small number as an index in a table called a hash table.
+
```
Comment (Bayler) [inline content/articles/hashing-in-data-structure/index.md:20]: Briefly introduce what is hashing in general and what it entails. Most readers of this articles are beginner level. It's always good to put that in mind when writing an article.

File: content/articles/hashing-in-data-structure/index.md (hunk 1 above)
Comment (Bayler) [reply content/articles/hashing-in-data-structure/index.md:24]: Okay. Noted. No need of adding any sources.

**Comment by Bayler** [comment]
@anubhavb11 Please work on the revisions then we can proceed to final review.

**Comment by anubhavb11** [comment]
@Bayler  I have worked on all the revisions you can have a look.

File: content/articles/hashing-in-data-structure/index.md [hunk 6]
```diff
@@ -0,0 +1,314 @@
...
+![output](./Capture.PNG)
+
+* In the output you can see we have inserted different key-value pairs into the custom hash table build by us. Try to experiment with these values and you will find collisions also and rehashing too if there are many collisions.
+
+* Searching happening in O(1) because of the hash table.
+
+#### Conclusion
+
+With the help of hash tables, we can insert, search and delete in O(1) time which is a great achievement. Hash tables are widely used for making our code more efficient and faster. Here are some problems of data structures you can try and think the approach using hash tables.
```
Comment (Bayler) [inline content/articles/hashing-in-data-structure/index.md:305]: Searching happens in 0(1) and 0(1) time? What do you mean by this?

Comment (anubhavb11) [reply content/articles/hashing-in-data-structure/index.md]: @Bayler  I have tried to make it more clear you can have a look.

**Comment by Bayler** [comment]
HI @anubhavb11 this article is now ready for final review.

**Comment by hectorkambow** [comment]
plagiarism check done

**Review by louisefindlay23 — COMMENTED** [review]
Congrats on publishing your first article, @anubhavb11. 

Please use a unique branch for each article in the future though i.e. hashing-data-structures rather than your forked repo's master branch.

Merged using a duplicate PR - cc @Bayler.
